Complete redesign of the resource registry

edsko opened the pull request.

mrBliss commented on ouroboros-consensus/tools/db-convert/Main.hs:
```diff
@@ -166,7 +166,7 @@ validateChainDb
   -> IO ()
 validateChainDb dbDir cfg onlyImmDB verbose =
-  ResourceRegistry.with $ \registry -> give protocolMagicId $ give epochSlots $
+  ResourceRegistry.withRegistry $ \registry -> give protocolMagicId $ give epochSlots $
```
Remove the qualifier, now we have `Registry` 3x on one line. It's almost as bad as Java 😀

edsko replied:
Yeah, fair enough. Just wanted to make minimal changes to stuff not directly in consensus. Will change.

mrBliss commented on ouroboros-consensus/test-util/Test/Util/SOP.hs:
```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE ScopedTypeVariables #-}
```
You know that there's somebody who would want you to make this change in a separate commit 🙂

edsko replied:
Thankfully that someone is on holidays :grin:

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/BlockchainTime.hs:
```diff
@@ -62,55 +63,51 @@ data BlockchainTime m = BlockchainTime {
+      -- Unlike the thread created by 'onSlotChange_', this thread will
+      -- terminate as soon as the action does.
+      --
+      -- Use sites should call 'onSlot_' rather than 'onSlot_'.
```
The first `onSlot_` should be `onSlot`

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:
```diff
@@ -96,7 +96,7 @@ run
 run tracers chainDbTracer rna dbPath pInfo
     customiseChainDbArgs customiseNodeArgs onNodeKernel =
-    ResourceRegistry.with $ \registry -> do
+    ResourceRegistry.withRegistry $ \registry -> do
```
I would remove the qualifier

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:
```diff
@@ -273,22 +273,26 @@ initNetwork
 initNetwork registry nodeArgs kernel RunNetworkArgs{..} = do
     -- serve local clients (including tx submission)
-    localServer <- forkLinked registry runLocalServer
+    localServer <- forkThread registry runLocalServer
```
Would it sense to add `forkLinkedThread`?

edsko replied:
It seemed unnecessary, but since both you and @nfrisby prefer it, I'll add it :)

edsko replied:
I'll remove the `forkThreadSelf`, I thought we'd need it in tests but turns out it doesn't help. And it avoids requiring a `forkLinkedThreadSelf` variant then too.

mrBliss commented on ouroboros-consensus/src/Ouroboros/Consensus/Node.hs:
```diff
@@ -273,22 +273,26 @@ initNetwork
+       linkToRegistry worker
+       return worker
 
     let asyncs = localServer : peerServer : ipSubscriptions : dnsSubscriptions
```
`threads` would be more accurate than `asyncs`

mrBliss commented on ouroboros-consensus/src/Ouroboros/Storage/ChainDB/Impl/Background.hs:
```diff
@@ -82,10 +82,15 @@ launchBgTasks cdb@CDB{..} = do
       copyToImmDBRunner cdb gcSchedule
     lgrSnapshotThread <- launch $
       updateLedgerSnapshotsRunner cdb
+    -- TODO: Do we still need this cdbBgThreads, if we have the registry?
```
Hmm, I think the only reason we have `cdbBgThreads` is to be able to shut down the background threads before we close the ImmutableDB and the VolatileDB, so that no background thread tries to access them after their closure, which would result in a `ClosedDBError`.

Since the ImmutableDB and the VolatileDB are not managed by the `ResourceRegistry` (maybe they should be?!), these actions do not take place in the right order, I believe.

edsko replied:
I'll leave it for now. It kinda feels to me that this `MVar` is a registry in disguise, and so maybe we should just use one, but it's not essential to this PR.

edsko commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:
```diff
@@ -1,324 +1,763 @@
+-- 'withRegistry', thereby deallocating @r@ -- leaving the thread to run with
+-- a now deallocated resource.
+--
+-- It is /only/ safe for threads to use a given registry if the lifetime of
```
Yup, all sounds good to me. Made the changes.

edsko commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:
```diff
@@ -1,324 +1,763 @@
+-- the scope of that registry, and hence the lifetime of the thread that created
+-- it. So, when we call 'linkToRegistry', the exception will be thrown the
+-- thread that created the registry, which (if not caught) will cause that that
+-- to exit the scope of 'withRegistry', thereby terminating the entire hierarchy
```
Addressing the first point: renamed `newRegistry` to `unsafeNewRegistry`. Also added a section "Combinators primarily for testing" in the exports, and move `unsafeNewRegistry`, `closeRegistry` as well as `countResources` to that section.

edsko replied:
As regards terminology, I'd rather we assume less rather than more, and be explicit where needed. I've added a note:

```
-- Note on terminology: when thread A forks thread B, we will say that thread A
-- is the " parent " and thread B is the " child ". No further relationship
-- between the two threads is implied by this terminology. In particular, note
-- that the child may outlive the parent. We will use "fork" and "spawn"
-- interchangeably.
```

I agree that "hierarchy" is indeed misleading, and I've rephrased the text to avoid it (details below).

edsko replied:
I've weakened the "chaining" paragraph to say:

```
-- Linking a parent to a child is however of limited value if the lifetime of
-- the child is not limited by the lifetime of the parent. For example, if A
-- does
--
-- > threadB <- async $ ..
-- > link threadB
--
-- and A terminates before B does, any exception thrown by B might be send to a
-- thread that no longer exists. This is particularly problematic when we start
-- chaining threads: if A spawns-and-links-to B which spawns-and-links-to C, and
-- C throws an exception, perhaps the intention is that this gets rethrown to B,
-- and then rethrown to A, terminating all three threads; however, if B has
-- terminated before the exception is thrown, C will throw the exception to a
-- non-existent thread and A is never notified.
```

edsko replied:
Avoiding another use of hierarchy:

```
-- thread that created the registry, which (if not caught) will cause that that
-- to exit the scope of 'withRegistry', thereby terminating all threads in that
-- registry.
```

edsko replied:
The last use of "hierachy" is justified:

```
-- This /also/ includes nested calls to 'withRegistry'. Since the lifetime of
-- such a registry (and all resources within) is tied to the thread calling
-- 'withRegistry', which itself is tied to the "parent registry" in which it was
-- created, this creates a hierarchy of registries.
```

edsko replied:
> Is that too thin of a scenario to accommodate without a concrete example to back it up? 

No, I don't think so, I thought about it also, but I guess I wasn't specific enough in the documentation. I've added the following part to the last section:

```
-- This includes 'withAsync': after all, 'withAsync' can't result in a child
-- thread that outlives its parent, so this is safe. In particular, if a parent
-- thread wants to link to a child thread /and handle any exceptions arising
-- from the link/, it should probably use `withAsync`:
--
-- > let handleLinkException :: ExceptionInLinkedThread -> m ()
-- >     handleLinkException = ..
-- > in handle handleLinkException $
-- >      withAsync codeInChild $ \child ->
-- >        ..
--
-- instead of
--
-- > handle handleLinkException $ do  -- PROBABLY NOT CORRECT!
-- >   child <- async codeInChild
-- >   ..
--
-- where the parent may exit the scope of the exception handler before the child
-- terminates. If the lifetime of the child cannot be limited to the lifetime of
-- the parent, the child should probably be linked to the registry instead and
-- the thread that spawned the registry should handle any exceptions.
```

edsko commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:
```diff
@@ -1,324 +1,791 @@
+          Left closed -> do
+            -- Despite the earlier check, it's possible that the registry got
+            -- closed after we allocated a new key but before we got a chance to
+            -- register the resource. In this case, we must deallocate the
```
Have left a comment about this in the code for now.

edsko commented on ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs:
```diff
@@ -1,324 +1,791 @@
+
+-- | Wait for thread to terminate and return its result.
+--
+-- If the thread throws an exception, this will rethrow that exception.
```
Good point. Added comment:

```
-- NOTE: If A waits on B, and B is linked to the registry, and B throws an
-- exception, then A might /either/ receive the exception thrown by B /or/
-- the 'ThreadKilled' exception thrown by the registry.
```

edsko commented:
Addressed the feedback from @mrBliss and @nfrisby . Moreover, did another thorough peer review over hangout with @nfrisby which resulted in a few subtle tweaks. I think this is ready for a final review by @nfrisby and can then be merged.
